pkg/services/ngalert/api/api.go

```diff
@@ -49,6 +50,27 @@ type Alertmanager interface {
 	TestReceivers(ctx context.Context, c apimodels.TestReceiversConfigBodyParams) (*notifier.TestReceiversResult, error)
 }
 
+type MultiOrgAlertmanager interface {
+	AlertmanagerFor(orgID int64) (Alertmanager, error)
+}
+
+type AlertmanagerProvider struct {
```

Comment:
This was necessary because `notifier.MultiOrgAlertmanager.AlertmanagerFor` returned a concrete `notifier.Alertmanager` which made it no longer conform to the interface above. This is because Go forces signatures to match exactly when applying interfaces. It's essentially this problem: https://stackoverflow.com/questions/34190170/go-interface-method-returning-interface-doesnt-match-method-returning-concrete

I worked around this by introducing this type, which adapts the concrete class's signature `AlertmanagerFor(orgID int64) (notifier.Alertmanager, error)` to return the interface instead, `AlertmanagerFor(orgID int64 (Alertmanager, error)`. Please let me know if there is a better way to do this.

Also [this](https://github.com/grafana/grafana/blob/6ae8f23f326193b907e7d4116447ecdfd5a8eef0/pkg/services/ngalert/api/api_alertmanager.go#L513) function already exists which does something similar. Another possibility would be to move `AlertmanagerProvider` to that file, use that as the adapter method instead.

Reply:
I think I would opt-in for the second option, reusing the existing function.

Reply:
I do not think I like adding more entities just for the sake of working around some language limitations. Perhaps we can figure out a better way, like a return interface instead of a concrete type. @gotjosh wdyt?

Reply:
@yuri-tceretian We could do this by pushing the interface down into the `notifier` package rather than keeping all the interfaces in the `api` package. This type can disappear entirely if `MultiOrgAlertmanager.AlertmanagerFor` returned an interface directly.

This was an approach I considered at first I avoided it originally because it felt more idiomatic in Go to keep the interface bound to the consuming layer. But, perhaps I followed this to a fault here - it definitely seems simpler to go with what you suggested.

Reply:
I ended up re-using the test infra in `notifier/multiorg_alertmanager_test.go` which let me re-use existing fakes for the persistence layer. This let me use a concrete alertmanager in the test, working around this problem.

IMO, this results in a less isolated/reproducible unit test, since it hits lots of unrelated functionality. But, as @JohnnyQQQQ mentioned offline, we'll be refactoring this early next year anyway.